### PR TITLE
Add TPS

### PR DESCRIPTION
**Pull request by 2793145003**
- Add TPS as a preprocessor
(But I'm not sure if I write the source correctly...)

### CODE REVIEW

File: mmocr/models/textrecog/preprocessor/tps_preprocessor.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
+# used code from https://github.com/clovaai/deep-text-recognition-benchmark/blob/master/modules/transformation.py
+
```
Comment (innerlee) [inline mmocr/models/textrecog/preprocessor/tps_preprocessor.py:2]: Add info of the code's licence

File: mmocr/models/textrecog/preprocessor/tps_preprocessor.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        """ Based on RARE TPS
+        input:
+            batch_I: Batch Input Image [batch_size x I_channel_num x I_height x I_width]
+            I_size : (height, width) of the input image I
+            I_r_size : (height, width) of the rectified image I_r
+            I_channel_num : the number of channels of the input image I
+        output:
+            batch_I_r: rectified image [batch_size x I_channel_num x I_r_height x I_r_width]
+        """
+        super(TPSPreprocessor, self).__init__()
```
Comment (innerlee) [inline mmocr/models/textrecog/preprocessor/tps_preprocessor.py:25]: `super().__init__()`

File: mmocr/models/textrecog/preprocessor/tps_preprocessor.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+import torch.nn.functional as F
+
+from mmocr.models.builder import PREPROCESSOR
+from .base_preprocessor import BasePreprocessor
+
+@PREPROCESSOR.register_module()
+class TPSPreprocessor(BasePreprocessor):
+    """ Rectification Network of RARE, namely TPS based STN """
+
+    def __init__(self, F, I_size, I_r_size, I_channel_num=1):
```
Comment (innerlee) [inline mmocr/models/textrecog/preprocessor/tps_preprocessor.py:15]: > F

Typically we do not use a single letter as variable name. Need to make their names more descriptive.

File: mmocr/models/textrecog/preprocessor/tps_preprocessor.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        """ Based on RARE TPS
+        input:
+            batch_I: Batch Input Image [batch_size x I_channel_num x I_height x I_width]
```
Comment (innerlee) [inline mmocr/models/textrecog/preprocessor/tps_preprocessor.py:18]: The docstring format should be adapted to our style (similar to pytorch's) :)

**Comment by cuhk-hbsun** [comment]
ci_lint is failed, please run `pre-commit run --all-files` before `git commit`

**Comment by innerlee** [comment]
The linting fails. Please run 
```bash
pre-commit install
pre-commt run --all-files
```
at the root of the repo

**Comment by 2793145003** [comment]
Thanks for the relpy.
Let me know if there's any other problem.

**Comment by innerlee** [comment]
The next step is to add some unittests. The goal is to cover the new code as fully as possible. You can check the coverage status from https://github.com/open-mmlab/mmocr/pull/117#issuecomment-824785474 and links therein.

File: mmocr/models/textrecog/preprocessor/tps_preprocessor.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+@PREPROCESSOR.register_module()
+class TPSPreprocessor(BasePreprocessor):
+    """Rectification Network of RARE, namely TPS based STN."""
+
+    def __init__(self, num_fiducial, I_size, I_r_size, I_channel_num=1):
+        """ Based on RARE TPS
+        Args:
+            num_fiducial: number of fiducial points of TPS-STN
+            I_size : (height, width) of the input image I
```
Comment (innerlee) [inline mmocr/models/textrecog/preprocessor/tps_preprocessor.py:32]: I_size -> img_size

File: mmocr/models/textrecog/preprocessor/tps_preprocessor.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+            I_r_size : (height, width) of the rectified image I_r
```
Comment (innerlee) [inline mmocr/models/textrecog/preprocessor/tps_preprocessor.py:33]: I_r_size -> rectified_img_size

File: mmocr/models/textrecog/preprocessor/tps_preprocessor.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+            I_channel_num : the number of channels of the input image I
```
Comment (innerlee) [inline mmocr/models/textrecog/preprocessor/tps_preprocessor.py:34]: I_channel_num -> img_channel

File: mmocr/models/textrecog/preprocessor/tps_preprocessor.py [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+        output:
+            batch_I_r: rectified image
+                [batch_size x I_channel_num x I_r_height x I_r_width]
+        """
+        super().__init__()
+        self.num_fiducial = num_fiducial
+        self.I_size = I_size
+        self.I_r_size = I_r_size  # = (I_r_height, I_r_width)
```
Comment (innerlee) [inline mmocr/models/textrecog/preprocessor/tps_preprocessor.py:42]: rm the comment since they are in docstring

**Comment by 2793145003** [comment]
well I tried. let me know if I write it wrong...

File: mmocr/models/textrecog/preprocessor/tps_preprocessor.py [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+    def init_weights(self):
+        pass
+
+
+class LocalizationNetwork(nn.Module):
+    """Localization Network of RARE, which predicts C' (K x 2) from input
+    (img_width x img_height)"""
+
+    def __init__(self, num_fiducial, num_img_channel):
+        super(LocalizationNetwork, self).__init__()
```
Comment (innerlee) [inline mmocr/models/textrecog/preprocessor/tps_preprocessor.py:83]: super().__init__()

File: mmocr/models/textrecog/preprocessor/tps_preprocessor.py [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+        return batch_C_prime
+
+
+class GridGenerator(nn.Module):
+    """Grid Generator of RARE, which produces P_prime by multipling T with
+    P."""
+
+    def __init__(self, num_fiducial, rectified_img_size):
+        """Generate P_hat and inv_delta_C for later."""
+        super(GridGenerator, self).__init__()
```
Comment (innerlee) [inline mmocr/models/textrecog/preprocessor/tps_preprocessor.py:149]: super().__init__()

**Comment by innerlee** [comment]
Thanks @2793145003 , we have taken a pass on coding styles. I will take another pass on the functionality.

Besides, since this is a model, a config containg this module is needed, so that users know how to use it. Some documentation on what it is, and how to use it will be great. We can offer help regarding these two

**Comment by 2793145003** [comment]
I'm wondering wether should I put the config file in crnn/ or in a new folder, since TPS is just a preprocessor and can be used in many models.
And I used TPS in my own dataset, so I don't have results that can be shown...

**Comment by innerlee** [comment]
A new folder is okay. If the config is modified from `crnn/crnn_academic_dataset.py`, then the new config could be `tps/crnn_tps_academic_dataset.py`.
The `configs/textrecog/crnn/README.md` could also be created to give the paper information.

**Comment by 2793145003** [comment]
I added the paper information, but it seems too crude, any advice?

**Comment by innerlee** [comment]
The paper info is okay. We will also have a section of results on academic datasets and provide training logs and checkpoints. Do you want to train one model on academic dataset, and report the results here? If not, don't worry, we will verify this later.

**Comment by 2793145003** [comment]
well, my gpu is being used for trying to get better result on my own dataset, maybe I can train on academic dataset after I finish it  but I'm not sure when I can do it...

**Review by cuhk-hbsun — APPROVED** [review]

**Comment by OpenMMLab-Coodinator** [comment]
Hi @2793145003 ！First of all, we want to express our gratitude for your significant PR in this project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA

If you have WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤

**Pull request merged**